File: Components/Strata.Foundation.Core/java/strata/foundation/core/container/IMultiMap.java
// ##########################################################################
// # File Name:	IMultiMap.java
// ##########################################################################

package strata.foundation.core.container;

import java.util.Collection;
import java.util.List;

public 
interface IMultiMap<K,V>
{
    IMultiMap<K,V>
    put(K key,V value);
    
    IMultiMap<K,V>
    remove(K key);
    
    IMultiMap<K,V>
    remove(K key,V value);
    
    Collection<V>
    get(K key);
    
    V
    get(K key,int index);
    
    int
    getCardinality(K key);
    
    Collection<K>
    getKeys();
    
    Collection<List<V>>
    getValues();
    
    boolean
    containsKey(K key);
    
    boolean
    containsValue(K key,V value);
}

// ##########################################################################

File: Components/Strata.Foundation.Core/main/typescript/Container/IMultiMap.ts
export
interface IMultiMap<K,V>
{
    put(key:K,value:V): IMultiMap<K,V>;

    remove(key:K): IMultiMap<K,V>;

    remove(key: K,value: V): IMultiMap<K,V>;

    get(key: K): Array<V>;

    getAt(key: K,index: number): V;

    getCardinality(key: K): number;

    getKeys(): Array<K>;

    getValues(): Array<Array<V>>;

    containsKey(key: K): boolean;

    containsValue(key: K,value: V): boolean;

}